### PR TITLE
CORS-2900: CAPI AWS IAM

### PR DESCRIPTION
**Pull request by patrickdillon**
Generate the IAM roles used by the control plane.

### CODE REVIEW

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to load AWS session: %w", err)
+	}
+	svc := iam.New(session)
+
+	// Create the IAM Roles for master and workers.
+	clusterOwnedIAMTag := &iam.Tag{
+		Key:   aws.String(fmt.Sprintf("kubernetes.io/cluster/%s", clusterID)),
+		Value: aws.String("owned"),
+	}
```
Comment (patrickdillon) [inline pkg/infrastructure/aws/clusterapi/iam.go:103]: need to add user tags

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:12]: Let's not add this dep :). You should use the stdlib `errors` for the `errors.As` calls.

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+					},
+					Resource: iamv1.Resources{"*"},
+				},
+			},
+		},
+	}
+)
+
+// putIAMRoles creates the roles used by control-plane and compute nodes.
+func putIAMRoles(clusterID string, ic *installconfig.InstallConfig) error {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:89]: Let's add a context to the argument list

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+		Value: aws.String("owned"),
+	}
+	assumePolicy := &iamv1.PolicyDocument{
+		Version: "2012-10-17",
+		Statement: iamv1.Statements{
+			{
+				Effect: "Allow",
+				Principal: iamv1.Principals{
+					iamv1.PrincipalService: []string{
+						"ec2.amazonaws.com",
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:111]: If we hardcode this value, it'll fail for C2S as it did for the AWS SDK (see fix in https://github.com/openshift/installer/pull/7871/files)

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+			// If the role does not exist, create it.
+			createRoleInput := &iam.CreateRoleInput{
+				RoleName:                 roleName,
+				AssumeRolePolicyDocument: aws.String(string(assumePolicyBytes)),
+				Tags:                     []*iam.Tag{clusterOwnedIAMTag},
+			}
+			if _, err := svc.CreateRole(createRoleInput); err != nil {
+				return fmt.Errorf("failed to create %s role: %w", role, err)
+			}
+			time.Sleep(10 * time.Second)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:141]: Is the sleep still need with the wait in the following line?

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		},
+	}
+	assumePolicyBytes, err := json.Marshal(assumePolicy)
+	if err != nil {
+		return fmt.Errorf("failed to marshal assume policy: %w", err)
+	}
+
+	for _, role := range []string{"master", "worker"} {
+		roleName := aws.String(fmt.Sprintf("%s-%s-role", clusterID, role))
+		if _, err := svc.GetRole(&iam.GetRoleInput{RoleName: roleName}); err != nil {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:127]: Let's use the `WithContext` version of this function: 
```suggestion
		if _, err := svc.GetRoleWithContext(ctx, &iam.GetRoleInput{RoleName: roleName}); err != nil {
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+			if errors.As(err, &awsErr) && awsErr.Code() != iam.ErrCodeNoSuchEntityException {
+				return fmt.Errorf("failed to get %s role: %w", role, err)
+			}
+			// If the role does not exist, create it.
+			createRoleInput := &iam.CreateRoleInput{
+				RoleName:                 roleName,
+				AssumeRolePolicyDocument: aws.String(string(assumePolicyBytes)),
+				Tags:                     []*iam.Tag{clusterOwnedIAMTag},
+			}
+			if _, err := svc.CreateRole(createRoleInput); err != nil {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:138]:
```suggestion
			if _, err := svc.CreateRoleWithContext(ctx, createRoleInput); err != nil {
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+				return fmt.Errorf("failed to create %s role: %w", role, err)
+			}
+			time.Sleep(10 * time.Second)
+			if err := svc.WaitUntilRoleExists(&iam.GetRoleInput{RoleName: roleName}); err != nil {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:142]:
```suggestion
			if err := svc.WaitUntilRoleExistsWithContext(ctx, &iam.GetRoleInput{RoleName: roleName}); err != nil {
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+			}
+		}
+
+		// Put the policy inline.
+		policyName := aws.String(fmt.Sprintf("%s-%s-policy", clusterID, role))
+		b, err := json.Marshal(policies[role])
+		if err != nil {
+			return fmt.Errorf("failed to marshal %s policy: %w", role, err)
+		}
+		if _, err := svc.PutRolePolicy(&iam.PutRolePolicyInput{
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:153]:
```suggestion
		if _, err := svc.PutRolePolicyWithContext(ctx, &iam.PutRolePolicyInput{
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+			PolicyDocument: aws.String(string(b)),
+			PolicyName:     policyName,
+			RoleName:       roleName,
+		}); err != nil {
+			return fmt.Errorf("failed to create inline policy for role %s: %w", role, err)
+		}
+
+		profileName := aws.String(fmt.Sprintf("%s-%s-profile", clusterID, role))
+		if _, err := svc.GetInstanceProfile(&iam.GetInstanceProfileInput{InstanceProfileName: profileName}); err != nil {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:162]:
```suggestion
		if _, err := svc.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{InstanceProfileName: profileName}); err != nil {
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+			if aerr, ok := err.(awserr.Error); ok && aerr.Code() != iam.ErrCodeNoSuchEntityException {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:163]: Please use the `errors.As` pattern.

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+				return fmt.Errorf("failed to get %s instance profile: %w", role, err)
+			}
+			// If the profile does not exist, create it.
+			if _, err := svc.CreateInstanceProfile(&iam.CreateInstanceProfileInput{
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:167]:
```suggestion
			if _, err := svc.CreateInstanceProfileWithContext(ctx, &iam.CreateInstanceProfileInput{
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+				InstanceProfileName: profileName,
+				Tags:                []*iam.Tag{clusterOwnedIAMTag},
+			}); err != nil {
+				return fmt.Errorf("failed to create %s instance profile: %w", role, err)
+			}
+			time.Sleep(10 * time.Second)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:173]: Ditto

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+			if err := svc.WaitUntilInstanceProfileExists(&iam.GetInstanceProfileInput{InstanceProfileName: profileName}); err != nil {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:174]:
```suggestion
			if err := svc.WaitUntilInstanceProfileExistsWithContext(ctx, &iam.GetInstanceProfileInput{InstanceProfileName: profileName}); err != nil {
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+				return fmt.Errorf("failed to wait for %s role to exist: %w", role, err)
+			}
+
+			// Finally, attach the role to the profile.
+			if _, err := svc.AddRoleToInstanceProfile(&iam.AddRoleToInstanceProfileInput{
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:179]:
```suggestion
			if _, err := svc.AddRoleToInstanceProfileWithContext(ctx, &iam.AddRoleToInstanceProfileInput{
```

File: pkg/infrastructure/aws/clusterapi/aws.go [hunk 16]
```diff
@@ -12,3 +15,10 @@ type Provider struct{}
 
 // Name gives the name of the provider, AWS.
 func (*Provider) Name() string { return awstypes.Name }
+
+func (*Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionInput) error {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/aws.go:19]: Should also add the check
```go
var _ clusterapi.PreProvider = (*Provider)(nil)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+			// If the profile does not exist, create it.
+			if _, err := svc.CreateInstanceProfile(&iam.CreateInstanceProfileInput{
+				InstanceProfileName: profileName,
+				Tags:                []*iam.Tag{clusterOwnedIAMTag},
+			}); err != nil {
+				return fmt.Errorf("failed to create %s instance profile: %w", role, err)
+			}
+			time.Sleep(10 * time.Second)
+			if err := svc.WaitUntilInstanceProfileExists(&iam.GetInstanceProfileInput{InstanceProfileName: profileName}); err != nil {
+				return fmt.Errorf("failed to wait for %s role to exist: %w", role, err)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:175]:
```suggestion
				return fmt.Errorf("failed to wait for %s instance profile to exist: %w", role, err)
```

**Comment by patrickdillon** [comment]
Feedback addressed and squashed into the initial commit.

Added bd6ebd260296bda4713406f8ca647c4cd955c1e3 for the BYO IAM role scenario.

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+func getOrCreateIAMRole(ctx context.Context, nodeRole, infraID, assumePolicy string, ic installconfig.InstallConfig, tags []*iam.Tag, svc *iam.IAM) (string, error) {
+	roleName := aws.String(fmt.Sprintf("%s-%s-role", infraID, nodeRole))
+
+	var defaultRole string
+	if dmp := ic.Config.AWS.DefaultMachinePlatform; dmp != nil && len(dmp.IAMRole) > 0 {
+		defaultRole = dmp.IAMRole
+	}
+
+	masterRole := defaultRole
+	if cp := ic.Config.ControlPlane.Platform.AWS; cp != nil && len(cp.IAMRole) > 0 {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:198]: Won't this fail if there is not `controlPlane` stanza defined in the installconfig? I think you need:
```suggestion
	if cp := ic.Config.ControlPlane; cp != nil && cp.Platform.AWS != nil && len(cp.Platform.AWS.IAMRole) > 0 {
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+					},
+					Resource: iamv1.Resources{"*"},
+				},
+			},
+		},
+	}
+)
+
+// putIAMRoles creates the roles used by control-plane and compute nodes.
+func putIAMRoles(ctx context.Context, infraID string, ic *installconfig.InstallConfig) error {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:94]: Nit:
```suggestion
// putIAMRoles creates the roles used by control-plane and compute nodes.
func createIAMRoles(ctx context.Context, infraID string, ic *installconfig.InstallConfig) error {
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+			},
+		},
+	}
+	assumePolicyBytes, err := json.Marshal(assumePolicy)
+	if err != nil {
+		return fmt.Errorf("failed to marshal assume policy: %w", err)
+	}
+
+	for _, role := range []string{master, worker} {
+		roleName, err := getOrCreateIAMRole(ctx, role, infraID, string(assumePolicyBytes), *ic, tags, svc)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:141]: `err` is not being checked here.

File: pkg/infrastructure/aws/clusterapi/iam.go (hunk 18 above)
Comment (patrickdillon) [reply pkg/infrastructure/aws/clusterapi/iam.go:198]: I wondered the same thing, but in my testing it didn't fail. So it seems like `ControlPlane` is always initialized.

That said, I still think it's good to check and will add it. :+1:

**Comment by patrickdillon** [comment]
Addressed feedback and tested both creating and BYO IAM role scenarios.

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Comment by r4f4** [comment]
/override ci/prow/altinfra-e2e-aws-capi-ovn
Expected to fail at this point.

**Pull request merged**